src/compiler/sourcemap.ts

```diff
@@ -577,6 +363,326 @@ namespace ts {
+            generatedCharacter = pendingGeneratedCharacter;
+
+            if (hasPendingSource) {
+                // 2. Relative sourceIndex
+                sourceMapData.sourceMapMappings += base64VLQFormatEncode(pendingSourceIndex - sourceIndex);
+                sourceIndex = pendingSourceIndex;
+
+                // 3. Relative source line
+                sourceMapData.sourceMapMappings += base64VLQFormatEncode(pendingSourceLine - sourceLine);
+                sourceLine = pendingSourceLine;
+
+                // 4. Relative source character
+                sourceMapData.sourceMapMappings += base64VLQFormatEncode(pendingSourceCharacter - sourceCharacter);
+                sourceCharacter = pendingSourceCharacter;
+
+                if (hasPendingName) {
+                    // 5. Relative nameIndex
+                    sourceMapData.sourceMapMappings += base64VLQFormatEncode(pendingNameIndex - nameIndex);
+                    nameIndex = pendingNameIndex;
+                }
+            }
+
+            hasMapping = true;
+            exit();
+        }
+
+        function toJSON(): RawSourceMap {
+            commitPendingMapping();
+            return {
+                version: 3,
+                file: sourceMapData.sourceMapFile,
+                sourceRoot: sourceMapData.sourceMapSourceRoot,
+                sources: sourceMapData.sourceMapSources,
+                names: sourceMapData.sourceMapNames,
+                mappings: sourceMapData.sourceMapMappings,
+                sourcesContent: sourceMapData.sourceMapSourcesContent,
+            };
+        }
+    }
+
+    function getSourceMapEmitter(generator: SourceMapGenerator, writer: EmitTextWriter): SourceMapEmitter {
+        if (writer.getSourceMapEmitter) {
+            return writer.getSourceMapEmitter(generator);
+        }
+
+        return createSourceMapEmitter(generator, writer);
+    }
+
+    function createSourceMapEmitter(generator: SourceMapGenerator, writer: EmitTextWriter) {
+        return {
+            emitMapping,
+            emitSourceMap
+        };
+
+        function emitMapping(sourceIndex?: number, sourceLine?: number, sourceCharacter?: number, nameIndex?: number) {
+            return generator.addMapping(writer.getLine(), writer.getColumn(), sourceIndex!, sourceLine!, sourceCharacter!, nameIndex);
+        }
+
+        function emitSourceMap(sourceMap: RawSourceMap, sourceMapPath: string): void {
+            return generator.appendSourceMap(writer.getLine(), writer.getColumn(), sourceMap, sourceMapPath);
+        }
+    }
+
+    function isStringOrNull(x: any) {
+        // tslint:disable-next-line:no-null-keyword
+        return typeof x === "string" || x === null;
+    }
+
+    export function isRawSourceMap(x: any): x is RawSourceMap {
+        // tslint:disable-next-line:no-null-keyword
+        return x !== null
+            && typeof x === "object"
+            && x.version === 3
+            && typeof x.file === "string"

```

Comment:
While `version` and `file` are technically required in the spec.... we don't use them, and we don't need them to be present to handle a map correctly. Do we really need to check for them?

Reply:
Its a small enough test on a non-hot code-path, and I'd prefer to align with the spec here.